### DISCUSSION TITLE
API Models Knowlege

### DISCUSSION THREAD

**Discussion by tomcom05** [open-webui/open-webui#4869] [Q&A]
I am having an odd issue. I have created a custom model and have added knowledge to it. When i chat with the model inside of the Open WebUI web interface the LLM responds with the knowledge i have added, but when i use the API its not recognizing the knowledge. 

Is this behavior expected? Should I be doing something in the model settings to include the knowledge during API calls? Hoping you all can help.

**Comment by mdlmarkham** [comment]
I'm very interested in this use case - how are you calling the custom model via the API? (I haven't been able to get that far yet ;-)

**Reply by tomcom05** [reply]
I am using the Extended OpenAI Conversation add-on in Home Assistant to make the call and then putting the name of the Model from Open WebUI into the field in the add-on. Seems to work except when trying to have the model utilize added knowledge.

**Comment by don-eg** [comment]
Is there a way to do what was asked here?  While the API appears to accept my custom model as a valid model, it does not include the RAG/Knowledge that I've configured in that model.

Seems like it would be useful to be able to simply configure and call the custom model that inherently uses the knowledge rather than having to pass the knowledge/collection in the API call.

**Reply by ahmetegesel** [reply]
I just came here to see if I am the only one who genuinely expected it to work like this. But I was so sad to see it didn't work. I hope it can be a real thing in the API. It would be super useful!

**Comment by RolandoQuintana** [comment]
Any updates on this from anyone??

**Comment by mahenning** [comment]
I tested it with a model with attached knowledge of a Nvidia yearly report and it works with a curl request to the API endpoint.
```
curl -X POST \
  https://<open-webui-address>/api/v1/chat/completions \
  -H "Authorization: Bearer sk-<key>" \
  -H "Content-Type: application/json" \
  -d '{
        "model": "testknowledge-bot",
        "messages": [
          {
            "role": "user",
            "content": "Whats the financial gain of Nvidia?"
          }
        ]
      }'
```
Answer: 
> "content": "According to the context, Nvidia's financial gain can be seen in several areas:\n\n* Revenue: Nvidia's revenue for fiscal year 2024 was $60.9 billion, up 126% from fiscal year 2023.\n* Net income: Nvidia's net income for fiscal year 2024 was $29.76 billion, up 581% from fiscal year 2023.\n* Gross margin: Nvidia's gross margin for fiscal year 2024 was 72.7%, up 15.8 points from fiscal year 2023.\n* Operating income: Nvidia's operating income for fiscal year 2024 was $32.97 billion, up 681% from fiscal year 2023.\n* Earnings per share: Nvidia's diluted earnings per share for fiscal year 2024 was $11.93, up 586% from fiscal year 2023.\n\nThese numbers indicate a significant financial gain for Nvidia in fiscal year 2024, driven by strong demand for their products, particularly in the data center and gaming markets."
<details><summary>Answer details (abridged)</summary>

```
{
    "sources": [
        {
            "source": {
                "id": "8ea6586e-a05c-46d1-bd60-ecd5bcceeca9",
                "user_id": "0168d81e-ebe3-4976-adee-70421e802f46",
                "name": "testknowledge",
                "description": "this is a test",
                "meta": null,
                "access_control": {
                    "read": {
                        "group_ids": []
                    },
                    "write": {
                        "group_ids": []
                    }
                },
                "created_at": <date>,
                "updated_at": <date>,
                "user": {
                    <user_data>
                },
                "files": [
                    {
                        "id": "cf1291cd-ce6d-4d58-9bc4-efd6b7354175",
                        "meta": {
                            "name": "2203.09507v3.pdf",
                            "content_type": "application/pdf",
                            "size": 4312542,
                            "collection_name": "8ea6586e-a05c-46d1-bd60-ecd5bcceeca9"
                        },
                        "created_at": 1741179736,
                        "updated_at": 1741179736
                    }
                ],
                "type": "collection"
            },
            "document": [
                <document sources>
            ],
            "metadata": [
                {
                    "page": 132,
                    "start_index": 0,
                    "hash": "9a77020530c47d141659b70cf78d90f677e5ff0b5a99741c5646e0a496898a09",
                    "created_by": "0168d81e-ebe3-4976-adee-70421e802f46",
                    "source": "NVIDIA-2024-Annual-Report.pdf",
                    "embedding_config": "{\"engine\": \"\", \"model\": \"intfloat/multilingual-e5-base\"}",
                    "file_id": "2ef2144e-9d0f-4550-8673-3c8fec4039a3",
                    "name": "NVIDIA-2024-Annual-Report.pdf",
                    "score": 0.9860798716545105
                },
				...
            ],
            "distances": [
                0.9860798716545105,
                0.9812244176864624,
                ...
            ]
        }
    ],
    "id": "chatcmpl-c66f4115a50442db8e0c749071206e38",
    "object": "chat.completion",
    "created": 1760513436,
    "model": "kosbu/Llama-3.3-70B-Instruct-AWQ",
    "choices": [
        {
            "index": 0,
            "message": {
                "role": "assistant",
                "content": "According to the context, Nvidia's financial gain can be seen in several areas:\n\n* Revenue: Nvidia's revenue for fiscal year 2024 was $60.9 billion, up 126% from fiscal year 2023.\n* Net income: Nvidia's net income for fiscal year 2024 was $29.76 billion, up 581% from fiscal year 2023.\n* Gross margin: Nvidia's gross margin for fiscal year 2024 was 72.7%, up 15.8 points from fiscal year 2023.\n* Operating income: Nvidia's operating income for fiscal year 2024 was $32.97 billion, up 681% from fiscal year 2023.\n* Earnings per share: Nvidia's diluted earnings per share for fiscal year 2024 was $11.93, up 586% from fiscal year 2023.\n\nThese numbers indicate a significant financial gain for Nvidia in fiscal year 2024, driven by strong demand for their products, particularly in the data center and gaming markets.",
                "refusal": null,
                "annotations": null,
                "audio": null,
                "function_call": null,
                "tool_calls": [],
                "reasoning_content": null
            },
            "logprobs": null,
            "finish_reason": "stop",
            "stop_reason": null,
            "token_ids": null
        }
    ],
    "service_tier": null,
    "system_fingerprint": null,
    "usage": {
        "prompt_tokens": 10241,
        "total_tokens": 10453,
        "completion_tokens": 212,
        "prompt_tokens_details": null
    },
    "prompt_logprobs": null,
    "prompt_token_ids": null,
    "kv_transfer_params": null
}

```
</details> 
Worked both at version 0.6.28 and 0.6.32. I left out the actual sources text and other stuff in the answer details, otherwise the text gets pretty huge. But you can see that it gets and uses sources from the files. So it looks like the problem is more the "Extended OpenAI Conversation add-on", and how it uses the API. But the `/v1/chat/completions` endpoint I used is the default chat entry point, there is nothing special about it.
